src/auth/src/token_cache.rs

```diff
@@ -0,0 +1,540 @@
+// TODO(#1552): Use the token cache in all creds
+#[allow(dead_code)]
+impl<T> TokenCache<T>
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    pub(crate) fn new(inner: T) -> Self {
+        let (tx_token, rx_token) = watch::channel::<Option<Result<Token>>>(None);
+
+        tokio::spawn(async move {
+            refresh_task(inner, tx_token).await;
+        });
+
+        Self {
+            rx_token,
+            _phantom: std::marker::PhantomData,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<T> TokenProvider for TokenCache<T>
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    async fn get_token(&self) -> Result<Token> {
+        let mut rx = self.rx_token.clone();
+        let token_result = rx.borrow_and_update().clone();
+
+        if let Some(token_result) = token_result {
+            match token_result {
+                Ok(token) => match token.expires_at {
+                    None => return Ok(token),
+                    Some(e) => {
+                        if e < Instant::now().into_std() {
+                            // Expired token, wait for refresh
+                            return wait_for_next_token(rx).await;
+                        } else {
+                            // valid token
+                            return Ok(token);
+                        }
+                    }
+                },
+                // An error in the result is still a valid result to propagate to the client library
+                Err(e) => Err(e),
+            }
+        } else {
+            wait_for_next_token(rx).await
+        }
+    }
+}
+
+async fn wait_for_next_token(
+    mut rx_token: watch::Receiver<Option<Result<Token>>>,
+) -> Result<Token> {
+    rx_token.changed().await.unwrap();
+    let token_result = rx_token.borrow().clone();
+
+    token_result.expect("There should always be a token or error in the channel after changed()")
+}
+
+async fn refresh_task<T>(token_provider: T, tx_token: watch::Sender<Option<Result<Token>>>)
+where
+    T: TokenProvider + Send + Sync + 'static,
+{
+    loop {
+        let token_result = token_provider.get_token().await;
+
+        let _ = tx_token.send(Some(token_result.clone()));
+
+        match token_result {
+            Ok(new_token) => {
+                if let Some(expiry) = new_token.expires_at {
+                    let time_until_expiry = expiry.duration_since(Instant::now().into_std());
+
+                    if time_until_expiry > FIVE_MINUTES {
+                        let time_to_sleep = Instant::now() + time_until_expiry - FIVE_MINUTES;
+                        sleep_until(time_to_sleep).await;
+                    } else if time_until_expiry > TEN_SECONDS {
+                        // If expiry is less than 5 mins, wait till last 10 seconds to refresh
+                        // This is to handle cases where MDS returns about to refresh tokens
+                        let time_to_sleep = Instant::now() + time_until_expiry - TEN_SECONDS;
```

Comment:
This is very odd, if we normally think that starting 5 minutes earlier is a good idea, then why wait until there are only 10 seconds left?  maybe we should just sleep for 10 seconds in this case?  Or maybe do some kind of exponentially shorter sleeps (1/2 of the remaining time) if the expiration is within 5 minutes?

Reply:
updated as per offline discussion.